src/katsdpmodels/primary_beam.py

```diff
@@ -0,0 +1,256 @@
+
+        The `target` must have ``obstime`` and ``obslocation`` properties, and
+        must be scalar. It will be converted to ICRS if necessary.
+        """
+        # TODO: implement
+        raise NotImplementedError
+
+
+class OutputType(enum.Enum):
+    JONES_HV = 1
+    JONES_XY = 2
+    MUELLER = 3
+    UNPOLARIZED_POWER = 4
+
+
+class Parameter:
+    """Description of a parameter accepted by a model.
+
+    Parameters
+    ----------
+    name
+        Name of the parameter used when passing it to sampling functions
+    description
+        Human-readable description of the parameter
+    unit
+        Units that must be used for the parameter
+    required
+        Whether the parameter is required to use the model
+    """
+
+    def __init__(self, name: str, description: str, unit: str, required: bool) -> None:
+        self.name = name
+        self.description = description
+        self.unit = unit
+        self.required = required
+
+
+class PrimaryBeam(models.SimpleHDF5Model):
+    r"""Base class for primary beams.
+
+    The phase of the electric field at a point increases with time i.e., the
+    phasor is
+
+    .. math:: e^{(\omega t - kz)i}
+
+    The "ideal" complex voltage is multiplied by the beam model to obtain the
+    measured voltage.
+
+    If you stick your right arm out and left hand up, then pretend to be an
+    antenna (lean back and look at the sky) then they correspond to the
+    directions of positive horizontal and vertical polarization respectively
+    (the absolute signs don't matter but the signs relative to each other do
+    for cross-hand terms).
+
+    An instance of this class describes either a single antenna or an average
+    of multiple antennas; to do antenna-specific modelling you will need a
+    separate instance of this class per antenna.
+    """
+
+    model_type: ClassVar[Literal['primary_beam']] = 'primary_beam'
+
+    def resolution(self, frequency: u.Quantity) -> float:
+        """Approximate spatial resolution of the model, in units of projected coordinates.
+
+        Sampling a grid at significantly higher resolution than this will have
+        diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    @property
+    def frequency_range(self) -> Tuple[u.Quantity, u.Quantity]:
+        """Minimum and maximum frequency covered by the model."""
+        raise NotImplementedError()
+
+    @property
+    def frequency_resolution(self) -> u.Quantity:
+        """Approximate frequency resolution of the model.
+
+        Sampling at significantly higher spectral resolution than this will
+        have diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    def min_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has full coverage."""
```

Comment:
Yes, I did have trouble with the names. Basically if you have a square model, there is distance from the centre to an edge, and distance from the centre to a corner. I'm not 100% sure if both are going to be useful.

Reply:
I guess the mathematically appropriate terms would be `inradius` and `circumradius`. At least you are forcing the user to look it up instead of guessing what it means :joy:

Reply:
I like that actually.

Reply:
👍🏾